hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElement.java

```diff
@@ -403,7 +403,7 @@ public void setOrigin(FromElement origin, boolean manyToMany) {
 				ASTUtil.appendSibling( origin, this );
 			}
 			else {
-				if ( !getWalker().isInFrom() && !getWalker().isInSelect() ) {
+				if ( !getWalker().isInFrom() && !getWalker().isInSelect() && !getWalker().isInEntityGraph()) {
```

Comment:
It's easier to see looking add at it in context: https://github.com/gbadner/hibernate-core/blob/HHH-9392-master/hibernate-core/src/main/java/org/hibernate/hql/internal/ast/tree/FromElement.java#L406 .

Without the additional check (&& !getWalker().isInEntityGraph()), the FromElement was being added as a child to the FromClause; to fix, the FromElement needed to be added as a child to the origin.

Really, isInEntityGraph() is a hacky way of getting around the fact that isInFrom() is false when processing the EntityGraph.

I considered adding a protected setter for HqlSqlBaseWalker.inFrom that HqlSqlWalker could use when processing the EntityGraph, but that seemed hackier.